lib/active_model/serializer/fragment_cache.rb

```diff
@@ -0,0 +1,99 @@
+module ActiveModel
+  class Serializer
+    class FragmentCache
+
+      attr_reader :serializer
+
+      def initialize(adapter, serializer, options, root)
+        @root       = root
+        @options    = options
+        @adapter    = adapter
+        @serializer = serializer
+      end
+
+      def fetch
+        klass = serializer.class
+        serializers = fragment_serializer(@serializer.object.class.name, klass)
+
+        cached_serializer = serializers[:cached].constantize.new(@serializer.object)
+        non_cached_serializer = serializers[:non_cached].constantize.new(@serializer.object)
+
+        cached_hash = @adapter.class.new(cached_serializer, @options).serializable_hash
+        non_cached_hash = @adapter.class.new(non_cached_serializer, @options).serializable_hash
+
+        if @serializer.root && @adapter.class == ActiveModel::Serializer::Adapter::JsonApi
+          cached_hash_root(cached_hash, non_cached_hash)
```

Comment:
If you're adding code specific for `JsonApi`, I think you should push that into the adapter. You could use polymorphism.

You could create `ActiveModel::Serializer::Adapter::FragmentCache` and `ActiveModel::Serializer::Adapter::JsonApi::FragmentCache`. Then here you could just do:

```ruby
adapter.fragment_cache(cached_hash, non_cached_hash)
```

This would also get rid of `cached_hash_root` from this class, specially given it's `JsonApi` specific.